### DISCUSSION TITLE
⚡️Possible ERC-721 `safe_mint` Gas Optimisation

### DISCUSSION THREAD

**Discussion by jaglinux** [pcaversaccio/snekmate#40] [Ideas]
Sorry for nitpick, I was going through the code and spotted gas saving optimization.
https://github.com/pcaversaccio/snekmate/blob/main/src/tokens/ERC721.vy#L546
Using token_id  for incrementing self._counter saves gas, even though self._counter qualifies for warm touch,
reading state variable is costly than reading from local variable.

We can further optimize the code. As per the comment in next line, we can use unsafe_add() since we don't expect token_ids to exceed 2**256. However testSafeMintOverflow() fails if we add unsafe_add.

Also, the comments needs to move before L546 ?

I have a draft PR at https://github.com/jaglinux/snekmate/commit/4ea35215486e36f1143c6a4b9cfc0344fbbd5e18

**Comment by pcaversaccio** [comment]
Correct - we can use `token_id` to increment the counter. Want to open a PR?

"we can use unsafe_add()" -> no since it's always possible that someone changes the upper bound of `_all_tokens`. I want to include the checked arithmetic here, so we don't offer a foot gun! That's why my overflow test throws ;-) - I'm testing for this possibility.

"The comments needs to move before L546" -> no since the comment relates to the line that could potentially overflow (in this case an `internal` function call).

**Reply by jaglinux** [reply]
Yes, I will create the PR.
